src/devices/Arduino/FirmataDevice.cs

```diff
@@ -0,0 +1,1234 @@
+    public delegate void DigitalPinValueChanged(int pin, PinValue newValue);
+
+    internal delegate void AnalogPinValueUpdated(int pin, uint rawValue);
+
+    /// <summary>
+    /// Low-level communication layer for the firmata protocol. Creates the binary command stream for the different commands and returns back results.
+    /// </summary>
+    internal sealed class FirmataDevice : IDisposable
+    {
+        private const byte FIRMATA_PROTOCOL_MAJOR_VERSION = 2;
+        private const byte FIRMATA_PROTOCOL_MINOR_VERSION = 5; // 2.5 works, but 2.6 is recommended
+        private const int FIRMATA_INIT_TIMEOUT_SECONDS = 2;
+        private static readonly TimeSpan DefaultReplyTimeout = TimeSpan.FromMilliseconds(500);
+        private static readonly TimeSpan ProgrammingTimeout = TimeSpan.FromMinutes(2);
+
+        private byte _firmwareVersionMajor;
+        private byte _firmwareVersionMinor;
+        private byte _actualFirmataProtocolMajorVersion;
+        private byte _actualFirmataProtocolMinorVersion;
+
+        private int _lastRequestId;
+
+        private string _firmwareName;
+        private Stream? _firmataStream;
+        private Thread? _inputThread;
+        private bool _inputThreadShouldExit;
+        private List<SupportedPinConfiguration> _supportedPinConfigurations;
+        private IList<byte> _lastResponse;
+        private List<PinValue> _lastPinValues;
+        private Dictionary<int, uint> _lastAnalogValues;
+        private object _lastPinValueLock;
+        private object _lastAnalogValueLock;
+        private object _synchronisationLock;
+        private Queue<byte> _dataQueue;
+
+        // Event used when waiting for answers (i.e. after requesting firmware version)
+        private AutoResetEvent _dataReceived;
+
+        public event DigitalPinValueChanged? DigitalPortValueUpdated;
+
+        public event AnalogPinValueUpdated? AnalogPinValueUpdated;
+
+        public event Action<string, Exception?>? OnError;
+
+        public FirmataDevice()
+        {
+            _firmwareVersionMajor = 0;
+            _firmwareVersionMinor = 0;
+            _firmataStream = null;
+            _inputThreadShouldExit = false;
+            _dataReceived = new AutoResetEvent(false);
+            _supportedPinConfigurations = new List<SupportedPinConfiguration>();
+            _synchronisationLock = new object();
+            _lastPinValues = new List<PinValue>();
+            _lastPinValueLock = new object();
+            _lastAnalogValues = new Dictionary<int, uint>();
+            _lastAnalogValueLock = new object();
+            _dataQueue = new Queue<byte>();
+            _lastResponse = new List<byte>();
+            _lastRequestId = 1;
+            _firmwareName = string.Empty;
+        }
+
+        internal List<SupportedPinConfiguration> PinConfigurations
+        {
+            get
+            {
+                return _supportedPinConfigurations;
+            }
+        }
+
+        public void Open(Stream stream)
```

Comment:
should this be part of the constructor or static Create method?

Reply:
I prefer to have these separated, for better testability (unit testing is simpler if ctors don't automatically start threads etc)